kernel/inferCumulativity.ml

```diff
@@ -172,9 +170,27 @@ let rec infer_fterm cv_pb infos variances hd stk =
   | FRel _ -> infer_stack infos variances stk
   | FInt _ -> infer_stack infos variances stk
   | FFloat _ -> infer_stack infos variances stk
-  | FFlex fl ->
-    let variances = infer_table_key (info_env (fst infos)) variances fl in
-    infer_stack infos variances stk
+  | FFlex Names.(RelKey _ | VarKey _ as fl) ->
+    (* We could try to lazily unfold but then we have to analyse the
+       universes in the bodies, not worth coding at least for now. *)
+    begin match unfold_ref_with_args (fst infos) (snd infos) fl stk with
+    | Some (hd,stk) -> infer_fterm cv_pb infos variances hd stk
+    | None -> infer_stack infos variances stk
+    end
+  | FFlex (Names.ConstKey con as fl) -> begin
+      let def = unfold_ref_with_args (fst infos) (snd infos) fl stk in
+      try
+        try
+        let ov = variances in
+        let variances = infer_constant (info_env (fst infos)) variances con in
+        let variances = infer_stack infos variances stk in
+        if Option.has_some def && variances != ov then raise Unfold else variances
+        with TrivialVariance when Option.has_some def -> raise Unfold
+      with BadVariance _ | Unfold as e ->
```

Comment:
I remember thinking that there should be a nicer way to make this work than this exception system but I don't remember what that way was supposed to be.